compiler/rustc_codegen_llvm/src/back/lto.rs

```diff
@@ -142,23 +145,26 @@ fn prepare_lto(
 }
 
 fn get_bitcode_slice_from_object_data(obj: &[u8]) -> Result<&[u8], LtoBitcodeFromRlib> {
-    let mut len = 0;
-    let data =
-        unsafe { llvm::LLVMRustGetBitcodeSliceFromObjectData(obj.as_ptr(), obj.len(), &mut len) };
-    if !data.is_null() {
-        assert!(len != 0);
-        let bc = unsafe { slice::from_raw_parts(data, len) };
-
-        // `bc` must be a sub-slice of `obj`.
-        assert!(obj.as_ptr() <= bc.as_ptr());
-        assert!(bc[bc.len()..bc.len()].as_ptr() <= obj[obj.len()..obj.len()].as_ptr());
-
-        Ok(bc)
-    } else {
-        assert!(len == 0);
-        Err(LtoBitcodeFromRlib {
-            llvm_err: llvm::last_error().unwrap_or_else(|| "unknown LLVM error".to_string()),
-        })
+    // The object crate doesn't understand bitcode files, but we can just sniff for the possible
+    // magic strings here and return the whole slice directly.
+    if obj.starts_with(b"\xDE\xC0\x17\x0B") || obj.starts_with(b"BC\xC0\xDE") {
+        return Ok(obj);
+    }
+    match object::read::File::parse(obj) {
+        Ok(f) => match f
+            .section_by_name(".llvmbc")
+            .or_else(|| f.section_by_name(".llvm.lto"))
+            .or_else(|| f.section_by_name("__LLVM,__bitcode"))
+            .or_else(|| f.section_by_name(".ipa"))
```

Comment:
Can we instead factor out a function that returns the correct section name and then look that up? That way things will stay in sync.